Add initial version for the YouLess integration

gjong opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

Added a new integration to support the YouLess devices.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/15271

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [x] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

gjong committed:
Add initial version for the YouLess integration

gjong committed:
Fix styling issues and unit test config flow

gjong committed:
Fix the requirements and remove unneeded state_attributes property from sensor

gjong committed:
Fix the requirements and remove unneeded state_attributes property from sensor

gjong committed:
Speed up test significantly

Kane610 commented on homeassistant/components/youless/__init__.py:
```diff
@@ -0,0 +1,53 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up youless from a config entry."""
+    try:
+        hass.data[DOMAIN][entry.entry_id] = YoulessAPI(entry.data[CONF_HOST])
```
You're doing IO in async context which is not allowed, it is also not recommended for a lib to do IO as part of the creation of the object

gjong replied:
I'll change this around a bit so that the IO does not happen in the integration initialization. In the config_flow a similar IO check is done to verify that the IP provided actually exists. Is this type OK.

Kane610 replied:
If you can wrap it in such a way that you can call it with async_add_executor_job it should be ok

gjong replied:
I've changed it so there is no longer a direct IO call from the async methods.

gjong committed:
Prevent direct IO calling from async operations

gjong committed:
Prevent direct IO calling from async operations

gjong committed:
Add test for unhappy flow config

gjong committed:
Remove unneeded catch

gjong committed:
Fix sensor type and typo in the name

gjong committed:
Fix sensor type and typo in the name

gjong committed:
Update library version underneath to support LS110

gjong committed:
Update library version underneath to support LS110

Bert-R commented:
Any chance this is going to land in the April release?

pdwonline commented:
I hope so. I do not really know why It didn't last month..


Best regards

Paul de Wit




> Op 23 mrt. 2021, om 08:50 heeft Bert Roos ***@***.***> het volgende geschreven:
> 
> 
> Any chance this is going to land in the April release?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/home-assistant/core/pull/41942#issuecomment-804693742>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABTWIPSP4OKGKZBGS465I6TTFBB4ZANCNFSM4STHZLVQ>.
>

Bert-R commented:
@Kane610 Do you happen to know?

frenck commented:
@Bert-R Please don't randomly ping people and pull them in. That is considered impolite in general. Thanks 👍

Bert-R commented:
I'm sorry. Given Robert is one of the reviewers, I thought me might know.

frenck commented on homeassistant/components/youless/__init__.py:
```diff
@@ -0,0 +1,56 @@
+from homeassistant.const import CONF_HOST
+from homeassistant.core import HomeAssistant
+
+from ...exceptions import PlatformNotReady
```
Only use relative paths within the current component. In this case, use a full import.

frenck commented on homeassistant/components/youless/__init__.py:
```diff
@@ -0,0 +1,56 @@
+    try:
+        await hass.async_add_executor_job(api.initialize)
+    except URLError as exception:
+        raise PlatformNotReady from exception
```
Raise `ConfigEntryNotReady` in this case.

frenck commented on homeassistant/components/youless/__init__.py:
```diff
@@ -0,0 +1,56 @@
+from ...exceptions import PlatformNotReady
+from .const import DOMAIN
+
+CONFIG_SCHEMA = vol.Schema({DOMAIN: vol.Schema({})}, extra=vol.ALLOW_EXTRA)
```
Unused? Remove it?

frenck commented on homeassistant/components/youless/__init__.py:
```diff
@@ -0,0 +1,56 @@
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
```
Return type missing.

```suggestion
async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```

frenck commented on homeassistant/components/youless/__init__.py:
```diff
@@ -0,0 +1,56 @@
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
```
Return type missing

```suggestion
async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```

frenck commented on homeassistant/components/youless/__init__.py:
```diff
@@ -0,0 +1,56 @@
+PLATFORMS = ["sensor"]
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
```
Return type missing

```suggestion
async def async_setup(hass: HomeAssistant, config: dict) -> bool:
```

frenck commented on homeassistant/components/youless/config_flow.py:
```diff
@@ -0,0 +1,62 @@
+DATA_SCHEMA = vol.Schema({vol.Required(CONF_NAME): str, vol.Required(CONF_HOST): str})
+
+
+async def validate_input(hass: core.HomeAssistant, data):
```
Missing typing for data and missing return type.

frenck commented on homeassistant/components/youless/config_flow.py:
```diff
@@ -0,0 +1,62 @@
+    try:
+        api = YoulessAPI(data[CONF_HOST])
+        await hass.async_add_executor_job(api.initialize)
+    except HTTPError as exception:
+        raise CannotConnect() from exception
+    except URLError as exception:
+        raise CannotConnect() from exception
```
Can be simplified:

```suggestion
    except (HTTPError, URLError) as exception:
        raise CannotConnect() from exception
```

Additionally:
 - Since this method is only called once, you might consider using the logic in place directly instead of a separate method.
 - There are no tests that covers the `HTTPError` case.

gjong replied:
I've moved the function into the `async_step_user` as advised. Good to know you can combine exceptions in one single catch in Python, I wasn't aware of that.

I will add additional tests for the case when an exception is thrown.

frenck commented on homeassistant/components/youless/config_flow.py:
```diff
@@ -0,0 +1,62 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_SCHEMA = vol.Schema({vol.Required(CONF_NAME): str, vol.Required(CONF_HOST): str})
```
What is name? Do we need to ask for that?

gjong replied:
The name is used for generating unique ids in the various sensors. The name is included in the id to enable the adding of multiple Youless devices with ids that make sense.

gjong replied:
I have changed the unique device ID to use the MAC address if available or a UUID from config_utils when unavailable.

The only reason to keep name is for displaying in the UI to identify which activated YouLess integration you are displaying. This is in the integrations screen, the name of the integration. I do not know of any other elegant way to identify the integration otherwise, since the devices do not expose any name. Especially for users that activate multiple instances of Youless because they have multiple devices.

frenck commented on homeassistant/components/youless/config_flow.py:
```diff
@@ -0,0 +1,62 @@
+                    title=info["title"],
+                    data={
+                        CONF_HOST: user_input[CONF_HOST],
+                        CONF_NAME: user_input[CONF_NAME],
+                        CONF_DEVICE: user_input[CONF_NAME],
```
Duplicate?

gjong replied:
True I am removing the duplicate

frenck commented on homeassistant/components/youless/manifest.json:
```diff
@@ -0,0 +1,16 @@
+  "requirements": [
+    "youless-api==0.7"
+  ],
+  "ssdp": [],
+  "zeroconf": [],
+  "homekit": {},
+  "dependencies": [],
```
Defaults can be removed.

```suggestion
```

frenck commented on homeassistant/components/youless/sensor.py:
```diff
@@ -0,0 +1,213 @@
+    """Initialize the integration."""
+    gateway = hass.data[DOMAIN][entry.entry_id]
+    device = entry.data[CONF_DEVICE]
+    sensors = [
+        GasSensor(gateway, device),
+        PowerMeterSensor(gateway, device, "low"),
+        PowerMeterSensor(gateway, device, "high"),
+        PowerMeterSensor(gateway, device, "total"),
+        CurrentPowerSensor(gateway, device),
+        DeliveryMeterSensor(gateway, device, "total"),
+        DeliveryMeterSensor(gateway, device, "current"),
+    ]
+
+    async_add_entities(sensors)
```
```suggestion
    async_add_entities(
        [
            GasSensor(gateway, device),
            PowerMeterSensor(gateway, device, "low"),
            PowerMeterSensor(gateway, device, "high"),
            PowerMeterSensor(gateway, device, "total"),
            CurrentPowerSensor(gateway, device),
            DeliveryMeterSensor(gateway, device, "total"),
            DeliveryMeterSensor(gateway, device, "current"),
        ]
    )
```

frenck commented on homeassistant/components/youless/sensor.py:
```diff
@@ -0,0 +1,213 @@
+    @property
+    def unique_id(self) -> Optional[str]:
+        """Return the uniquely generated id."""
+        return f"{DOMAIN}_{self._device}_{self._sensor_id}"
```
The device is a user-configured string, and thus not viable for use with a unique ID.

<https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-requirements>

gjong replied:
I will have to look into if I can get some type of unique id from the Youless device, for the LS120 that should not be a problem as this exposes its MAC address using an API. For the LS110 this might be more of a challenge since this device does not.

frenck commented on homeassistant/components/youless/sensor.py:
```diff
@@ -0,0 +1,213 @@
+        self._device = device
+        self._sensor_id = sensor_id
+
+    @Throttle(timedelta(seconds=30))
```
All sensors use the same update... and the throttle is now needed to prevent them from all doing the same thing.

Instead; use the `DataUpdateCoordinator`, see: <https://developers.home-assistant.io/docs/integration_fetching_data/#coordinated-single-api-poll-for-data-for-all-entities>

gjong replied:
I've added the use of the DataUpdateCoordinator. I hope I implemented it correctly.

frenck commented on homeassistant/components/youless/sensor.py:
```diff
@@ -0,0 +1,213 @@
+    async_add_entities(sensors)
+
+
+class YoulessBaseSensor(Entity):
```
Please rebase the PR on the latest `dev`, and use `SensorEntity` instead.

frenck commented on homeassistant/components/youless/sensor.py:
```diff
@@ -0,0 +1,213 @@
+        """Get the sensor for providing the value."""
+        return self._gateway.current_power_usage
+
+    @property
+    def device_class(self) -> Optional[str]:
+        """Return the class of this device, from component DEVICE_CLASSES."""
+        return DEVICE_CLASS_POWER
+
+    @property
+    def icon(self) -> Optional[str]:
+        """Return the icon to use in the frontend, if any."""
+        return "mdi:power-socket"
```
When a device class is provided, one should not provide an icon. It is the one or the other (more cases in this PR).

frenck commented on homeassistant/components/youless/strings.json:
```diff
@@ -0,0 +1,21 @@
+{
+  "title": "YouLess",
```
This is a brand name, I don't think it needs to be translated?

frenck commented on homeassistant/components/youless/strings.json:
```diff
@@ -0,0 +1,21 @@
+    },
+    "error": {
+      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]",
+      "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]",
```
Seems like `invalid_auth` isn't used?

frenck commented on homeassistant/components/youless/strings.json:
```diff
@@ -0,0 +1,21 @@
+    "error": {
+      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]",
+      "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]",
+      "unknown": "[%key:common::config_flow::error::unknown%]"
```
Seems like `unknown` isn't used?

frenck commented on homeassistant/components/youless/strings.json:
```diff
@@ -0,0 +1,21 @@
+      "unknown": "[%key:common::config_flow::error::unknown%]"
+    },
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
```
Seems like `already_configured` isn't used?

frenck commented on tests/components/youless/test_config_flows.py:
```diff
@@ -0,0 +1,68 @@
+            {"host": "localhost", "name": "YouLess Sensor"},
+        )
+
+    assert result2["type"] == "create_entry"
```
Assertions are missing that validate the data in the config entry are as expected.

gjong commented:
Thanks for the usefull review frenk. I will try and get the remarks fixed somewhere this week.

gjong committed:
Integrate some of the review pointers

gjong commented:
There appears to be a serious issue in the unit test for the config_flow. I don't have enough time right now to fully investigate. I will try to look in to this in the near future. Right now I'm setting this pull-request back to draft.

frenck committed:
Fix config flow tests

frenck committed:
Format manifest add iot_class

frenck committed:
Tweak and type config flow

frenck committed:
Use helpers in modukle init

frenck committed:
Remove unused setup method

frenck committed:
Move coordinator, cleanup sensor

frenck commented:
Hi there @gjong 👋 

I've seen the struggle and opened up the PR to look at it a bit, and went it bit overboard.
I hope you don't mind! If you do, please let me know I'll revert the changes at that point.

Things I did:
- Rebased the PR
- Fix the tests
- Added typing
- Making use of some new helpers and methods that reduce code
- Moved the data coordinator into the init method
- Adjusted the tests and config flow a bit

I hope this helps with moving forward on your contribution ❤️ 

If you need any help or have more questions, please let me know!

../Frenck

frenck commented:
One thing that still needs to be solved, is the unique ID part.

If the mac address is None, we have to solve it differently.

Maybe in that case we can fall back to the `entry.entry_id` of the config entry (which is allowed).

Veldkornet commented:
FYI, the URL `'http://' + host + '/d'` should respond the the model and mac as below:
`{"model":"LS120","mac":"xx:xx:xx:xx:xx:xx"}`

gjong commented:
Hi @frenck.

Thanks for your changes to the code base, off course I don't mind that is what open source is all about. I haven't had much time to look into this pull-request lately (since other open source projects required my attention). I am going to look into the last modifications you suggested this weekend and hope that I can finish the pull request then.

gjong committed:
Remove the name from the flow and prevent issues for LS110 (no mac)

gjong commented:
I have updated the pull request with the following:

- Removed the name from the config flow
- Set the entry_id in place of the mac when the latter is None (this could occur only in case of LS110 device)

gjong commented:
@frenck I have spotted something strange in the behaviour of the integration. I have had it running since the update in my home assistant but it stopped fetching updates this morning. It appears as if the data coordinator stopped working.

As soon as I reload the integration in the configuration section the data polling resumes. Is this something known and if so how can I trigger the coordinator to resume the polling in the integration.

frenck commented on homeassistant/components/youless/config_flow.py:
```diff
@@ -0,0 +1,50 @@
+                errors["base"] = "cannot_connect"
+            else:
+                return self.async_create_entry(
+                    title=DOMAIN,
```
Small suggestion, if one sets up multiple, it would be recognizable.
```suggestion
                    title=user_input[CONF_HOST],
```

frenck reviewed (commented):
One small comment, otherwise it looks good to me 👍

gjong committed:
Update homeassistant/components/youless/config_flow.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

gjong commented:
> > It appears as if the data coordinator stopped working.
> 
> The coordinator is used by many many integrations, I have never seen it stop to be honest

Strange, I can only see the end result in the UI. In the sense that it displays stail data, last update couple of days ago. So I assumed the coordinator might have some logic to stop polling if something out of the ordinary happens. Like stated when reloading the polling resumes. 

We could leave this as is and wait to see if other users also experience this issue, but at least in my system I've noticed this happening.

frenck commented:
> So I assumed the coordinator might have some logic to stop polling if something out of the ordinary happens.

Unless the API blocks the requests... even at that point, the coordinator would just keep polling with the set rate until it responds. The update logic for the coordinator is robust enough to can anything wrong and just try again later.

gjong committed:
Fix unit test + upgrade library version

gjong commented:
@frenck I've fixed the tests after the autocommit of your suggestion broke the assertion. I've also updated the library underneath. It did not include a time-out in the request to the remote device, so that may have caused issues with the coordinator. In the sense that it might be permanently blocking the request, as without time-out the python docs indicate it will wait indefinitely.

frenck commented:
Sounds like this PR is ready to get out of draft? :)

frenck commented on homeassistant/components/youless/translations/en.json:
```diff
@@ -0,0 +1,22 @@
+                }
+            }
+        }
+    },
+    "title": "youless"
```
```suggestion
    }
```

frenck reviewed (commented):
The language file seems out of sync, feel free to get this PR out of draft when this last change is in.

gjong commented:
I will update the language files. I intend to keep this on draft for a couple of days. Hopefully I can verify the polling issue to be gone with the update of the library underneath.

I don't want to release when the changes are great that people might experience this behaviour.

gjong committed:
Update homeassistant/components/youless/translations/en.json

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

frenck commented:
Alright, that is fine. We have a beta cut upcoming Wednesday for the August release, something to keep in mind (after that we have a week of beta, which does allow for fixes of course).

gjong commented:
The integration seems to bo working stable since the update so I moved it out of draft.

frenck reviewed (approved):
Look good to me 👍 

For follow-up PRs, you could look at:

- <https://developers.home-assistant.io/blog/2021/05/25/sensor_attributes/>
- Adding strict typing to the integration
- Tests for the sensors

MartinHjelmare commented on homeassistant/components/youless/__init__.py:
```diff
@@ -0,0 +1,58 @@
+        _LOGGER,
+        name="youless_gateway",
+        update_method=async_update_data,
+        update_interval=timedelta(seconds=2),
```
We normally don't allow a polling interval of less than 5 seconds. This is extra important here since the library doesn't support asyncio and we're using the executor thread pool when fetching new data.

MartinHjelmare commented on homeassistant/components/youless/config_flow.py:
```diff
@@ -0,0 +1,50 @@
+        errors = {}
+        if user_input is not None:
+            try:
+                api = YoulessAPI(user_input[CONF_HOST])
```
Please only wrap the line that can raise in the try... except block.

MartinHjelmare commented on homeassistant/components/youless/config_flow.py:
```diff
@@ -0,0 +1,50 @@
+                api = YoulessAPI(user_input[CONF_HOST])
+                await self.hass.async_add_executor_job(api.initialize)
+            except (HTTPError, URLError):
+                _LOGGER.exception("Cannot connect to host")
```
We only include the stack trace for unknown errors, not when we know the error type.

MartinHjelmare commented on homeassistant/components/youless/sensor.py:
```diff
@@ -0,0 +1,197 @@
+    )
+
+
+class YoulessBaseSensor(CoordinatorEntity, Entity):
```
We don't need to inherit `Entity` when inheriting `CoordinatorEntity`.

Please also inherit `SensorEntity` imported from the sensor integration.

MartinHjelmare commented on homeassistant/components/youless/sensor.py:
```diff
@@ -0,0 +1,197 @@
+        self._device_group = device_group
+        self._sensor_id = sensor_id
+
+        self._attr_unique_id = f"{DOMAIN}_{device}_{sensor_id}"
```
We don't need to include the integration domain in the unique id. The entity registry is aware of both the integration domain and the platform domain.

MartinHjelmare commented on homeassistant/components/youless/sensor.py:
```diff
@@ -0,0 +1,197 @@
+        if self.get_sensor is None:
+            return None
+
+        return self.get_sensor.unit_of_measurement
```
Please map the devices unit to Home Assistant defined common constants and return the Home Assistant constant here. Eg:
https://github.com/home-assistant/core/blob/debcc6689f22bf564ab08ee261aed6231ec88d3b/homeassistant/const.py#L402

MartinHjelmare commented on homeassistant/components/youless/strings.json:
```diff
@@ -0,0 +1,15 @@
+    "step": {
+      "user": {
+        "data": {
+          "name": "[%key:common::config_flow::data::name%]",
```
The name is not included in the config flow form.

MartinHjelmare commented on tests/components/youless/test_config_flows.py:
```diff
@@ -0,0 +1,72 @@
+    with patch(
+        "homeassistant.components.youless.config_flow.YoulessAPI",
+        return_value=mock_youless,
+    ) as mocked_youless:
```
We normally patch the integration setup function to avoid setting up the whole integration on all create entry results. See the scaffold test template for example.

MartinHjelmare commented on tests/components/youless/test_config_flows.py:
```diff
@@ -0,0 +1,72 @@
+        )
+
+    assert result2.get("type") == RESULT_TYPE_FORM
+    assert len(mocked_youless.mock_calls) == 1
```
Please assert the result error type too.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!
